File: src/binding/julia/openPMD.jl/src/Mesh.jl
# Mesh

@doc """
    @enum Geometry begin
        cartesian
        theta_mode
        cylindrical
        spherical
        other
    end
""" Geometry
export Geometry, cartesian, theta_mode, cylindrical, spherical, other

@doc """
    @enum DataOrder begin
        C
        F
    end
""" DataOrder
export DataOrder, C, F

@doc """
    mutable struct Mesh <: Container{MeshRecordComponent}
        ...
    end
""" Mesh
export Mesh

@doc """
    unit_dimension(mesh::Mesh)::SVector{7,Double}
""" unit_dimension
export unit_dimension

@doc """
    isscalar(mesh::Mesh)::Bool
""" isscalar
export isscalar

@doc """
    geometry(mesh::Mesh)::Geometry
""" geometry
export geometry

@doc """
    set_geometry!(mesh::Mesh, geom::Geometry)
""" set_geometry!
export set_geometry!

@doc """
    geometry_parameters(mesh::Mesh)::AbstractString
""" geometry_parameters
export geometry_parameters

@doc """
    set_geometry_parameters!(mesh::Mesh, params::AbstractString)
""" set_geometry_parameters!
export set_geometry_parameters!

@doc """
    data_order(mesh::Mesh)
""" data_order
export data_order

@doc """
    set_data_order!(mesh::Mesh, order::DataOrder)
""" set_data_order!
export set_data_order!

@doc """
    axis_labels(mesh::Mesh)::AbstractVector{<:AbstractString}
""" axis_labels
export axis_labels

@doc """
    set_axis_labels!(mesh::Mesh, labels::AbstractVector{<:AbstractString})
""" set_axis_labels!
export set_axis_labels!
@cxxdereference function set_axis_labels!(mesh::Mesh, labels::AbstractVector{<:AbstractString})
    return set_axis_labels1!(mesh, wrap_vector(labels))
end

@doc """
    grid_spacing(mesh::Mesh)::AbstractVector{CxxDouble}
""" grid_spacing
export grid_spacing

@doc """
    set_grid_spacing!(mesh::Mesh, spacing::AbstractVector{CxxDouble})
""" set_grid_spacing!
export set_grid_spacing!
@cxxdereference set_grid_spacing!(mesh::Mesh, spacing::AbstractVector{CxxDouble}) = set_grid_spacing1!(mesh, wrap_vector(spacing))

@doc """
    grid_global_offset(mesh::Mesh)::AbstractVector{CxxDouble}
""" grid_global_offset
export grid_global_offset

@doc """
    set_grid_global_offset!(mesh::Mesh, offset::AbstractVector{CxxDouble})
""" set_grid_global_offset!
export set_grid_global_offset!
@cxxdereference function set_grid_global_offset!(mesh::Mesh, offset::AbstractVector{CxxDouble})
    return set_grid_global_offset1!(mesh, wrap_vector(offset))
end

@doc """
    grid_unit_SI(mesh::Mesh)::CxxDouble
""" grid_unit_SI
export grid_unit_SI

@doc """
    set_grid_unit_SI!(mesh::Mesh, unit::CxxDouble)
""" set_grid_unit_SI!
export set_grid_unit_SI!

@doc """
    set_unit_dimension!(mesh::Mesh, unit_dim::Dict{UnitDimension,<:AbstractFloat})
""" set_unit_dimension!
export set_unit_dimension!
@cxxdereference function set_unit_dimension!(mesh::Mesh, unit_dim::Dict{UnitDimension,<:AbstractFloat})
    return set_unit_dimension1!(mesh,
                                SVector{7,Float64}(get(unit_dim, L, 0.0), get(unit_dim, M, 0.0), get(unit_dim, T, 0.0),
                                                   get(unit_dim, I, 0.0), get(unit_dim, θ, 0.0), get(unit_dim, N, 0.0),
                                                   get(unit_dim, J, 0.0)))
end

@doc """
    time_offset(mesh::Mesh)::CxxDouble
""" time_offset
export time_offset

@doc """
    set_time_offset!(mesh::Mesh, unit::CxxDouble)
""" set_time_offset!
export set_time_offset!

"""
    const SCALAR::AbstractString
"""
const SCALAR = SCALAR1()

File: src/binding/julia/Mesh.cpp
// Mesh

#include "defs.hpp"

// Define supertype relationships
namespace jlcxx
{
template <>
struct SuperType<Mesh>
{
    using type = Container<MeshRecordComponent>;
};
} // namespace jlcxx

void define_julia_Mesh(jlcxx::Module &mod)
{

    // Mesh::Geometry
    mod.add_bits<Mesh::Geometry>("Geometry", jlcxx::julia_type("CppEnum"));
    jlcxx::stl::apply_stl<Mesh::Geometry>(mod);

    mod.set_const("GEOMETRY_cartesian", Mesh::Geometry::cartesian);
    mod.set_const("GEOMETRY_theta_mode", Mesh::Geometry::thetaMode);
    mod.set_const("GEOMETRY_cylindrical", Mesh::Geometry::cylindrical);
    mod.set_const("GEOMETRY_spherical", Mesh::Geometry::spherical);
    mod.set_const("GEOMETRY_other", Mesh::Geometry::other);

    // Mesh::DataOrder
    mod.add_bits<Mesh::DataOrder>("DataOrder", jlcxx::julia_type("CppEnum"));
    jlcxx::stl::apply_stl<Mesh::DataOrder>(mod);

    mod.set_const("DATAORDER_C", Mesh::DataOrder::C);
    mod.set_const("DATAORDER_F", Mesh::DataOrder::F);

    // Mesh
    auto type = mod.add_type<Mesh>(
        "CXX_Mesh",
        // We don't wrap BaseRecord<T> for simplicity. We thus need to declare
        // Container<MeshRecordComponent> as our supertype.
        jlcxx::julia_base_type<Container<MeshRecordComponent>>());

    // These two functions come from our superclass
    // BaseRecord<MeshRecordComponent>. We declare them as if they were our own.
    type.method("cxx_unit_dimension", &Mesh::unitDimension);
    type.method("cxx_isscalar", &Mesh::scalar);

    type.method("cxx_geometry", &Mesh::geometry);
    type.method(
        "cxx_set_geometry!",
        static_cast<Mesh &(Mesh::*)(Mesh::Geometry g)>(&Mesh::setGeometry));
    type.method("cxx_geometry_parameters", &Mesh::geometryParameters);
    type.method("cxx_set_geometry_parameters!", &Mesh::setGeometryParameters);
    type.method("cxx_data_order", &Mesh::dataOrder);
    type.method("cxx_set_data_order!", &Mesh::setDataOrder);
    type.method("cxx_axis_labels", &Mesh::axisLabels);
    type.method("cxx_set_axis_labels!", &Mesh::setAxisLabels);
    type.method("cxx_grid_spacing", &Mesh::gridSpacing<double>);
    type.method("cxx_set_grid_spacing!", &Mesh::setGridSpacing<double>);
    type.method("cxx_grid_global_offset", &Mesh::gridGlobalOffset);
    type.method("cxx_set_grid_global_offset!", &Mesh::setGridGlobalOffset);
    type.method("cxx_grid_unit_SI", &Mesh::gridUnitSI);
    type.method("cxx_set_grid_unit_SI!", &Mesh::setGridUnitSI);
    type.method(
        "cxx_set_unit_dimension!",
        [](Mesh &mesh, const array_double_7 &unitDimension) {
            return mesh.setUnitDimension(std::map<UnitDimension, double>{
                {UnitDimension::L, unitDimension[uint8_t(UnitDimension::L)]},
                {UnitDimension::M, unitDimension[uint8_t(UnitDimension::M)]},
                {UnitDimension::T, unitDimension[uint8_t(UnitDimension::T)]},
                {UnitDimension::I, unitDimension[uint8_t(UnitDimension::I)]},
                {UnitDimension::theta,
                 unitDimension[uint8_t(UnitDimension::theta)]},
                {UnitDimension::N, unitDimension[uint8_t(UnitDimension::N)]},
                {UnitDimension::J, unitDimension[uint8_t(UnitDimension::J)]},
            });
        });
    type.method("cxx_time_offset", &Mesh::timeOffset<double>);
    type.method("cxx_set_time_offset!", &Mesh::setTimeOffset<double>);
}
